### PR TITLE
[ISSUE-744]: CIS hardening support

### PR DESCRIPTION
**Pull request by Aber4Nod**
### Background:
Currently CSI Baremetal components cannot be deployed and operate with CIS hardening support enabled, see more here: https://docs.rke2.io/security/hardening_guide.
Long story short, without CIS profiling (as it is in current installations) RKE2 will put an unrestricted policy in place that allows Kubernetes to run as though the PodSecurityPolicy admission controller was not enabled.
When ran with the CIS profile, RKE2 will put a much more restrictive set of policies in place. These policies meet such requirements as MustRunAsNonRoot (i.e. container could not run as root process without special roles/rolebindings).

### Problem Solving
To solve the current problem we should use the following primitives: 
1. SecurityContext settings on pod/container levels in case of running containers as nonrootusers.
2. Binding ServiceAccounts to specific PodSecurityPolicies in case of accessing host resources.
3. Using privileged settings at pod/container levels for special cases (e.g. for shared mount point types - bidirectional mount propagation in Kubernetes terms).

### Deep insight in RCA and it's solving for specific components:
* Components, without privileged requirements: Controller, NodeController, Operator. These components fail to start due to the following error:
  ```
  atlantic    7s          Warning   Failed              pod/csi-baremetal-node-controller-6f4b665fdc-jq6pg    Error: container has runAsNonRoot and image has non-numeric user (bmcsi), cannot verify user is non-root (pod: "csi-baremetal-node-controller-6f4b665fdc-jq6pg_atlantic(55de0246-47a1-413a-815c-3f603ba2b27f)", container: node-controller)
  ```
  This happens due to non numeric user specification for containers' start processes. Because CIS profiling restricts containers' process to be run as non root, it cannot determine which UID has the specified user. The most common way is SecurityContext restrictions - just specify the needed user ID and optionally specify runAsNonRoot: false setting:
  ```
  securityContext:
    enable: {{ .Values.$component.securityContext.enable }}
    runAsNonRoot: {{ .Values.$component.securityContext.runAsNonRoot }}
    runAsUser: {{ .Values.$component.securityContext.runAsUser }}   
  ```
* Components, which need access to **only** host level resources: SchedulerExtender, SchedulerExtenderPatcher. These components fail to start due to the following errors:
  ```
  atlantic      2s          Warning   FailedCreate        daemonset/csi-baremetal-se                     Error creating: pods "csi-baremetal-se-" is forbidden: PodSecurityPolicy: unable to admit pod: [spec.securityContext.hostNetwork: Invalid value: true: Host network is not allowed to be used spec.containers[0].hostPort: Invalid value: 8787: Host port 8787 is not allowed to be used. Allowed ports: [] spec.containers[0].hostPort: Invalid value: 8889: Host port 8889 is not allowed to be used. Allowed ports: []]
  atlantic      2s          Warning   FailedCreate        daemonset/csi-baremetal-se-patcher             Error creating: pods "csi-baremetal-se-patcher-" is forbidden: PodSecurityPolicy: unable to admit pod: [spec.volumes[1]: Invalid value: "hostPath": hostPath volumes are not allowed to be used spec.volumes[2]: Invalid value: "hostPath": hostPath volumes are not allowed to be used]
  ```
  To get access to the host level resources, the corresponding pod's ServiceAccounts (csi-node-sa, csi-baremetal-extender-sa) should be bound to specific PodSecurityPolicies (in the most common case - to privileged), which give access to specific host paths and host ports.
* Components, which need **only** privileged access: Node. This component should be bound to privileged PodSecurityPolicy and also specified as privileged at SecurityContext Container's setting. The need is conditioned by mount propagations: shared mount points (bidirectional) is only allowed for privileged containers.

## PR checklist
- [x] Add link to the issue
- [x] Choose Project
- [x] Choose PR label
- [ ] New unit tests added
- [ ] Modified code has meaningful comments
- [ ] All TODOs are linked with the issues
- [ ] All comments are resolved

### Settings Preparation
* Set SecurityContext for CSi Baremetal Operator deployment:
  ```
  securityContext:
    enable: true 
    runAsNonRoot: true
    runAsUser: 1000
  ```
* Set the following SecurityContexts at deployments.csi-baremetal.dell.com for Controller, NodeController components:
  ```
  securityContext:
    enable: true
    runAsNonRoot: true
    runAsUser: 1000
  ```
* Create (if not existed) privileged PodSecurityPolicy:
  ```
  hostname1:~ # kubectl get podsecuritypolicy privileged
  NAME         PRIV   CAPS   SELINUX    RUNASUSER   FSGROUP    SUPGROUP   READONLYROOTFS   VOLUMES
  privileged   true   *      RunAsAny   RunAsAny    RunAsAny   RunAsAny   false            *
  ```
  Create Role for this policy usage:
  ```
  apiVersion: rbac.authorization.k8s.io/v1
  kind: Role
  metadata:
    name: manager-role
    namespace: atlantic
  rules:
  - apiGroups:
    - policy
    resourceNames:
    - privileged
    resources:
    - podsecuritypolicies
    verbs:
    - use
  ``` 
  Create RoleBinding for this policy usage:
  ```
  apiVersion: rbac.authorization.k8s.io/v1
  kind: RoleBinding
  metadata:
    name: manager-rolebinding
    namespace: atlantic
  roleRef:
    apiGroup: rbac.authorization.k8s.io
    kind: Role
    name: manager-role
  subjects:
  - kind: ServiceAccount
    name: csi-node-sa
    namespace: atlantic
  - kind: ServiceAccount
    name: csi-baremetal-extender-sa
    namespace: atlantic
  ```
* Set the following SecurityContexts at deployments.csi-baremetal.dell.com for Node, SchedulerExtender, SchedulerExtenderPatcher components:
  ```
  securityContext:
    enable: true
    privileged: true
  ```
* Enable podSecurityPolicy at deployments.csi-baremetal.dell.com:
   ```
   podSecurityPolicy:
     enable: true
     resourceName: privileged
   ```
### Components Deployment:
* Deploy CSI Baremetal Operator and check it is Running:
   ```
   hostname1:~ # kubectl get po -n atlantic
   NAME                                             READY   STATUS    RESTARTS   AGE
   csi-baremetal-operator-758ccb6944-c7nns          1/1     Running   0          106m
   ```
* Deploy CSI Baremetal Driver and check it is Running:
   ```
   hostname1:~ # kubectl get po -n atlantic
   NAME                                             READY   STATUS    RESTARTS   AGE
   csi-baremetal-controller-8cfcccd99-lm6mq         4/4     Running   0          4d22h
   csi-baremetal-node-controller-5b455557bb-8p6bs   1/1     Running   0          4h33m
   csi-baremetal-node-pnswq                         4/4     Running   0          104m
   csi-baremetal-operator-758ccb6944-c7nns          1/1     Running   0          106m
   csi-baremetal-se-patcher-djsdm                   1/1     Running   0          105m
   csi-baremetal-se-tbw6m                           1/1     Running   0          106m
   ```

### CODE REVIEW

File: docs/README.md [hunk 1]
```diff
@@ -117,6 +117,95 @@ Installation process
       helm install csi-baremetal csi/csi-baremetal-deployment --set driver.drivemgr.type=halmgr \
       --set global.registry=$REGISTRY --set global.registrySecret=$DOCKER_REGISTRY_SECRET
       ```
+Feature Supporting
+------
+### CIS hardening
+For supporting CIS Hardening we need to add the following settings while installing CSI and CSI Baremetal Operator:
```
Comment (mishoyama) [inline docs/README.md:123]: Why can't we make this enabled by default?

File: docs/README.md [hunk 2]
```diff
@@ -117,6 +117,95 @@ Installation process
...
+* While installing CSI Baremetal Operator set SecurityContext:
+  ```
+  --set securityContext.enable=true --set securityContext.runAsNonRoot=true --set securityContext.runAsUser=1000
```
Comment (mishoyama) [inline docs/README.md:126]: What does `1000` mean?

File: docs/README.md [hunk 3]
```diff
@@ -117,6 +117,95 @@ Installation process
...
+    --set driver.securityContext.securityContext.enable=true --set driver.securityContext.securityContext.runAsNonRoot=true --set driver.securityContext.securityContext.runAsUser=1000
+    ```
+  * Create (if not existed) privileged PodSecurityPolicy:
+    ```
+    apiVersion: policy/v1beta1
+    kind: PodSecurityPolicy
+    metadata:
+      name: privileged
+      annotations:
+        seccomp.security.alpha.kubernetes.io/allowedProfileNames: '*'
```
Comment (mishoyama) [inline docs/README.md:141]: Is this feature in alpha state?

File: docs/README.md [hunk 4]
```diff
@@ -117,6 +117,95 @@ Installation process
...
+    spec:
+      privileged: true
+      allowPrivilegeEscalation: true
+      allowedCapabilities:
+      - '*'
+      volumes:
+      - '*'
+      hostNetwork: true
```
Comment (mishoyama) [inline docs/README.md:149]: `hostNetwork` is required for scheduler extender only. why do we need to grand this access to other components as well?

File: docs/README.md [hunk 5]
```diff
@@ -117,6 +117,95 @@ Installation process
...
+      fsGroup:
+        rule: 'RunAsAny'
+    ```
+  * Create Role for this policy usage:
+    ```
+    apiVersion: rbac.authorization.k8s.io/v1
+    kind: Role
+    metadata:
+      name: manager-role
+      namespace: atlantic
```
Comment (mishoyama) [inline docs/README.md:170]: `atantic` namespace?

**Comment by mishoyama** [comment]
@Aber4Nod which version of CIS hardening did you test? Is it applicable for RKE2 only?

File: docs/README.md (hunk 1 above)
Comment (Aber4Nod) [reply docs/README.md:123]: We can, enabled them

File: docs/README.md (hunk 2 above)
Comment (Aber4Nod) [reply docs/README.md:126]: It's just uid for container process (it's the same for the root namespace as well as the container namespace) and used in most cases when container process has no need in files modifications (as in our case). 
Generally there can be any uid here except 0 - in this case PSP admission controller will fail pod creation operation with error about root UID forbiddance. 
So, entrypoint will have also the same uid, as we state here, as owner respectively.

File: docs/README.md (hunk 3 above)
Comment (Aber4Nod) [reply docs/README.md:141]: This feature is about to get out of alpha, but I didn't found this type of annotation in beta. Also, I just copied this definition from default psp privileged specification.

File: docs/README.md (hunk 5 above)
Comment (Aber4Nod) [reply docs/README.md:170]: yes, it was tested at platform workloads as was stated at the issue

**Comment by Aber4Nod** [comment]
> @Aber4Nod which version of CIS hardening did you test? Is it applicable for RKE2 only?

1.6 as was stated at the issue. Hardening can also be done in other platforms, but in another way, so this changes will be suitable for them

File: charts/csi-baremetal-deployment/values.yaml [hunk 6]
```diff
@@ -50,6 +52,10 @@ driver:
       requests:
         cpu:
         memory:
+    securityContext:
```
Comment (rkiselenko) [inline charts/csi-baremetal-deployment/values.yaml:55]: Maybe it's better to have this as default values? I mean denote it in the CRDs

Comment (Aber4Nod) [reply charts/csi-baremetal-deployment/values.yaml:53]: Added default values for SecurityContext fields

File: docs/README.md (hunk 2 above)
Comment (mishoyama) [reply docs/README.md:127]: Why do we need to ask user to set this parameter? Do we need to changes things like this in our docker files?
```
RUN addgroup -S bmcsi && adduser -S bmcsi -G bmcsi

USER bmcsi
```

File: docs/README.md (hunk 5 above)
Comment (mishoyama) [reply docs/README.md:171]: namespace must be configured.

**Comment by mishoyama** [comment]
> > @Aber4Nod which version of CIS hardening did you test? Is it applicable for RKE2 only?
> 
> 1.6 as was stated at the issue. Hardening can also be done in other platforms, but in another way, so this changes will be suitable for them

We need 

> > @Aber4Nod which version of CIS hardening did you test? Is it applicable for RKE2 only?
> 
> 1.6 as was stated at the issue. Hardening can also be done in other platforms, but in another way, so this changes will be suitable for them

Sorry, I don't see any references here https://github.com/dell/csi-baremetal/issues/744 to 1.6. If you addressing 1.6 only it should be clearly stated here

File: charts/csi-baremetal-deployment/templates/csi-baremetal_v1_deployment.yaml [hunk 7]
```diff
@@ -32,6 +32,12 @@ spec:
       log:
         format: {{ .Values.driver.log.format }}
         level: {{ .Values.driver.log.level }}
+      {{- if .Values.driver.controller.securityContext.enable }}
+      securityContext:
```
Comment (mishoyama) [inline charts/csi-baremetal-deployment/templates/csi-baremetal_v1_deployment.yaml:36]: Do we really need to extend API with these parameters? Can we just make them by default or modify docker files?

**Comment by mishoyama** [comment]
Found that `PodSecurityPolicy` is going to be deprecated https://kubernetes.io/docs/reference/using-api/deprecation-guide/#psp-v125

File: docs/README.md (hunk 2 above)
Comment (Aber4Nod) [reply docs/README.md:127]: The user already exists. I described the problem above

File: docs/README.md (hunk 5 above)
Comment (Aber4Nod) [reply docs/README.md:171]: Don't understand the comment - this is an example of usage. What is implied by configuration here?

**Comment by Aber4Nod** [comment]
> > > @Aber4Nod which version of CIS hardening did you test? Is it applicable for RKE2 only?
> > 
> > 
> > 1.6 as was stated at the issue. Hardening can also be done in other platforms, but in another way, so this changes will be suitable for them
> 
> We need
> 
> > > @Aber4Nod which version of CIS hardening did you test? Is it applicable for RKE2 only?
> > 
> > 
> > 1.6 as was stated at the issue. Hardening can also be done in other platforms, but in another way, so this changes will be suitable for them
> 
> Sorry, I don't see any references here [dell/csi-baremetal#744](https://github.com/dell/csi-baremetal/issues/744) to 1.6. If you addressing 1.6 only it should be clearly stated here

updated the issue with this

File: charts/csi-baremetal-deployment/templates/csi-baremetal_v1_deployment.yaml (hunk 7 above)
Comment (Aber4Nod) [reply charts/csi-baremetal-deployment/templates/csi-baremetal_v1_deployment.yaml:36]: In my opinion it's much more better to configure the behavior, which are initiated by external behavior (e.g. enabling cis hardening).
In a lot of examples at k8scluster-automation - other components uses the same approach.

**Comment by Aber4Nod** [comment]
> Found that `PodSecurityPolicy` is going to be deprecated https://kubernetes.io/docs/reference/using-api/deprecation-guide/#psp-v125

I doubt that this will affect rke case. For vanilla we can make the separate issue with this - and support 1.25 case separately.

**Review by rkiselenko — APPROVED** [review]

File: pkg/scheduler_extender.go [hunk 8]
```diff
@@ -39,34 +35,32 @@ const (
...
-	EventRecorder events.EventRecorder
-	MatchPolicies []rbacv1.PolicyRule
+	PodSecurityPolicyVerifier          securityverifier.SecurityVerifier
+	SecurityContextConstraintsVerifier securityverifier.SecurityVerifier
 }
 
 // Update updates csi-baremetal-se or creates if not found
 func (n *SchedulerExtender) Update(ctx context.Context, csi *csibaremetalv1.Deployment, scheme *runtime.Scheme) error {
-	// in case of Openshift deployment and non default namespace - validate extender service accounts security bindings
+	// in case of Openshift deployment and non default namespace - validate node service accounts security bindings
```
Comment (safronovD) [inline pkg/scheduler_extender.go:44]: typo - node instead of extender

File: pkg/scheduler_extender.go [hunk 9]
```diff
@@ -39,34 +35,32 @@ const (
...
+			var verifierError securityverifier.Error
+			err = n.SecurityContextConstraintsVerifier.HandleError(ctx, csi, csi.Spec.Scheduler.ServiceAccount, err)
+			if errors.As(err, &verifierError) {
+				return nil
+			}
+			return err
+		}
+	}
+
+	// in case of podSecurityPolicy feature enabled - validate node service accounts security bindings
```
Comment (safronovD) [inline pkg/scheduler_extender.go:56]: typo - node instead of extender

File: charts/csi-baremetal-deployment/values.yaml [hunk 10]
```diff
@@ -162,6 +171,12 @@ driver:
 scheduler:
   enable: true
   serviceAccount: csi-baremetal-extender-sa
+  securityContext:
```
Comment (safronovD) [inline charts/csi-baremetal-deployment/values.yaml:174]: Do we use the same policies for extender and patcher?

**Review by safronovD — APPROVED** [review]

File: docs/README.md (hunk 2 above)
Comment (mishoyama) [reply docs/README.md:127]: 1. No need to user to choose user ID
2. Should be aligned with existing user in Docker files.

**Review by mishoyama — APPROVED** [review]

**Pull request merged**